keps/sig-node/20200203-memory-manager.md

```diff
@@ -0,0 +1,374 @@
+
+Memory Table is proposed to represent amount of pre-reserved and allocatable memory for certain memory type on a NUMA node. The table consists of Pre Reserved Zone and Allocatable Zone. The following figure shows the structure of Memory Map briefly.
+
+```
++-------------------+------------+--------------+
+|                   |     Allocatable Zone      |
++ Pre-reserved Zone +------------+--------------+
+|                   | Guaranteed |     Free     |
++-------------------+------------+--------------+
+```
+_Figure: The construction of Memory Table._
+
+The size of thwo zones(Pre-reserved Zone and Allocatable Zone) are determined during initialize process and those sizes are immutable values in runtime.
+- Pre Reserved Zone indicates certain amount of memory is pre-reserved for system such as kernel, OS demon, core components of kubernetes. The size of this zone is specified on the configuration flag `--pre-reserved-zone`. For more details see [The configuration of Memory Manager](#the-configuration-of-memory-manager) section.
+
+- Allocatable Zone indicates the total size of allocatable memory for the containers in Guaranteed QoS Pod. The size of this zone is calculated automatically by subtraction `the size of Pre Reserved Zone in a MemoryTable` from `total memory size of the MemoryTable`.
+
+- Guaranteed memory section of Allocatable Zone indicates total size of reserved memory for the containers in Guaranteed QoS Pod. The Guaranteed memory of certain container is taken from Free memory section.
+
+- Free memory section of Allocatable Zone indicates free memory that memory manager refers to calculate NUMA node affinity. When the container ,which acquired Guaranteed memory, is terminated, Guaranteed memory will be returned to Free memory.
+
+NOTE:
+- The number of memory types on the node varies by the configuration of pre-allocated hugepages on the node.
+
+#### Computing NUMA node affinity
+
+Memory Manager calculates topology hint of conventional memory and each hugepage size for the containers in Guaranteed QoS Pod. The topology hint is also known as NUMA node affinity which represents the NUMA node that has enough capacity of memory for requested memory type. 
+
+Memory Manager refers `Memory Map` to find out all NUMA nodes which has available memory for given memory type. Then it returns found NUMA nodes as topology hints. Here each topology hint maps to each NUMA node.
+
+- Here is brief example for two socket machine.
+  - A Pod requests 2Gi of memory and 1Gi of hugepages-1Gi.
+  - Memory Manager found NUMA node 0 has enough memory and hugepage but NUAM node 1 doesn't so that calculated NUMA node affinity is NUMA node 0.
+  - Memory Manager returns the topology hint 
+
+#### The configuration of Memory Manager
+
+The Node Allocatable memory of Node Allocatable Feature is calculated by following fomula:
+- [Node-Allocatable] = [Node-Capacity] - [Kube-Reserved] - [System-Reserved] - [Eviction-Threshold]
```

Comment:
Should we also take into account pods that already started with `Guaranteed` QoS and was pinned to the specific NUMA node?